src/sage/rings/species.py

```diff
@@ -0,0 +1,2337 @@
+            components = G.disjoint_direct_product_decomposition()
+            elm = self.one()
+            for component in components:
+                gens = [[cyc for cyc in gen.cycle_tuples() if cyc[0] in component]
+                        for gen in G.gens()]
+                H = PermutationGroup([gen for gen in gens if gen],
+                                     domain=component)
+                pi_H = {k: [e for e in v if e in component]
+                        for k, v in pi.items()}
+                a = self._indices(H, pi_H, check=check)
+                elm *= self.gen(a)
+            return elm
+
+        try:
+            X, a = G
+        except TypeError:
+            raise ValueError(f"{G} must be a permutation group or a pair specifying a group action on the given domain pi={pi}")
+        dompart = [sorted(pi.get(s, [])) for s in range(self._arity)]
+        S = PermutationGroup([tuple(b) for b in dompart if len(b) > 2]
+                             + [(b[0], b[1]) for b in dompart if len(b) > 1],
+                             domain=list(chain(*dompart)))
+        H = _stabilizer_subgroups(S, X, a)
+        if len(H) > 1:
+            raise ValueError("Action is not transitive")
+        return self(H[0], pi, check=check)
+
+    def grading_set(self):
+        r"""
+        Return the set of non-negative integer vectors, whose length is
+        the arity of ``self``.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import MolecularSpecies
+            sage: MolecularSpecies(["X", "Y"]).grading_set()
+            Integer vectors of length 2
+        """
+        return IntegerVectors(length=self._arity)
+
+    def subset(self, size):
+        """
+        Return the set of molecular species with given total cardinality.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import MolecularSpecies
+            sage: M = MolecularSpecies(["X", "Y"])
+            sage: M.subset(3)  # random
+            {X*E_2(Y), X*Y^2, C_3(Y), E_3(X), Y^3, Y*E_2(Y), C_3(X), X^2*Y,
+             E_3(Y), E_2(X)*Y, X*E_2(X), X^3}
+        """
+        result = Set()
+        for grade in IntegerVectors(size, length=self._arity):
+            result = result.union(self.graded_component(grade))
+        return result
+
+    def graded_component(self, mc):
+        """
+        Return the set of molecular species with given multicardinality.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import MolecularSpecies
+            sage: M = MolecularSpecies(["X", "Y"])
+            sage: M.graded_component([3, 2])  # random
+            {E_3(X)*Y^2, X^3*Y^2, X*E_2(X)*E_2(Y), X^3*E_2(Y),
+             {((1,2,3), (1,3)(4,5)): ({1, 2, 3}, {4, 5})},
+             X*{((1,2)(3,4),): ({1, 2}, {3, 4})}, X*E_2(X)*Y^2, E_3(X)*E_2(Y),
+             C_3(X)*Y^2, C_3(X)*E_2(Y)}
+        """
+        assert len(mc) == self._arity
```

Comment:
You shouldn't use `assert` to validate input (they should be tests that are expected to always pass). It is better to raise a `ValueError` instead.